bndtools.core/src/bndtools/central/Central.java

```diff
@@ -204,39 +203,52 @@ public static Workspace getWorkspace() throws Exception {
             if (ws != null) { // early check for workspace
                 return ws;
             }
-            try {
-                Workspace.setDriver(Constants.BNDDRIVER_ECLIPSE);
-                Workspace.addGestalt(Constants.GESTALT_INTERACTIVE, new Attrs());
+            ws = initializeWorkspace();
+            // The workspace has been initialized fully, set the field now
+            workspace = ws;
+        }
+        workspaceQueue.resolve(ws); // notify onWorkspaceInit callbacks
```

Comment:
You cannot change the invariant that a Deferred/Promise can only be resolved once. You would need to replace the Deferred with a new one. And you need to make sure onWorkspaceInit work across any change. You will now have the workspace/workspaceQueue variables paired and thus you need a critical section to access either of them since they must change together. This then makes things much more complex. This is why changing the code from them being set only once to being reset is hard. Is there not a better solution to your issue which does not require them to be reset?

Also, there is a bug with the changes to add onWorkspaceInit which I have raised which may obviate the whole thing anyway :-) #1430.